File: app/src/main/java/com/route/fragment/FirstFragment.java
package com.route.fragment;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;

import com.route.apis.ApiManager;
import com.route.modal.RoutesDocuments;
import com.route.routeme.JavaBarCodeActivity;
import com.route.routeme.R;
import com.route.routeme.databinding.FragmentFirstBinding;
import com.route.viewmodel.RoutesDataViewModel;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

public class FirstFragment extends Fragment {

    private static final String TAG = "GenAuth";

    protected final CompositeDisposable mDisposable = new CompositeDisposable();
    private FragmentFirstBinding binding;
    private RoutesDataViewModel model;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentFirstBinding.inflate(inflater, container, false);
        return binding.getRoot();

    }

    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        /*binding.buttonFirst.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                NavHostFragment.findNavController(FirstFragment.this)
                        .navigate(R.id.action_FirstFragment_to_SecondFragment);
            }
        });*/


        binding.barcodeBtn.setOnClickListener(v->{
            Intent intent = new Intent(getActivity(), JavaBarCodeActivity.class);
            someActivityResultLauncher.launch(intent);
        });


        model = new ViewModelProvider(requireActivity()).get(RoutesDataViewModel.class);
        model.clearDisposable();

        binding.databaseBtn.setOnClickListener(btnView -> {
            model.loadDatabaseList();
        });

        binding.collectionBtn.setOnClickListener(btnView -> {
            model.loadRoutesMeDataCollection();
        });

        binding.documentBtn.setOnClickListener(btnView -> {
            //model.loadRoutesDocument();
            model.loadRoutesDocumentJson();
        });

        model.getRoutesDocument().observe(requireActivity(), routesData -> {
            // update UI
            List<RoutesDocuments> routes = routesData.getDocuments();
            binding.resText.setText(routes.get(0).id);
        });

        model.getRoutesDocumentJson().observe(requireActivity(), routesJsonData -> {
            // update UI
            if(binding != null && binding.resText != null) {
                binding.resText.setText(routesJsonData.toString());
            }
        });

        model.getCollection().observe(requireActivity(), collectionJson -> {
            // update UI
            binding.resText.setText(collectionJson.toString());
        });

        model.getDatabaseList().observe(requireActivity(), databaseJson -> {
            // update UI
            binding.resText.setText(databaseJson.toString());
        });

        model.getRoutesError().observe(requireActivity(), error -> {
            binding.resText.setText(error);
        });

        String requiredValue = "http://routemeapp.com/TestCode01";
        NavController navController = NavHostFragment.findNavController(FirstFragment.this);
        Bundle bundle = new Bundle();
        bundle.putString("Url", requiredValue);
        DocumentFragment.s_URL = requiredValue;
        navController.setGraph(R.navigation.nav_graph, bundle);
        navController.navigate(R.id.action_FirstFragment_to_DocumentFragment);
        binding.title.setText("List of Routes");

    }

    @Override
    public void onResume() {
        super.onResume();
        binding.title.setText("Scan The Bar Code");
    }

    ActivityResultLauncher<Intent> someActivityResultLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            new ActivityResultCallback<ActivityResult>() {
                @Override
                public void onActivityResult(ActivityResult result) {
                    if (result.getResultCode() == Activity.RESULT_OK) {
                        // Here, no request code
                        Intent data = result.getData();
                        String requiredValue = data.getStringExtra("key");
                        Log.i("Response", ""+requiredValue);

                        NavController navController = NavHostFragment.findNavController(FirstFragment.this);
                        Bundle bundle = new Bundle();
                        bundle.putString("Url", requiredValue);
                        DocumentFragment.s_URL = requiredValue;
                        navController.setGraph(R.navigation.nav_graph, bundle);
                        navController.navigate(R.id.action_FirstFragment_to_DocumentFragment);
                        binding.title.setText("List of Routes");
                    }
                }});


    @Override
    public void onDestroyView() {
        mDisposable.clear();
        mDisposable.dispose();
        super.onDestroyView();
        binding = null;
    }

}

File: app/src/main/java/com/route/fragment/FirstFragment.kt
package com.route.fragment

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.activity.OnBackPressedCallback
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.NavHostFragment
import com.google.android.material.snackbar.Snackbar
import com.google.gson.JsonElement
import com.route.modal.RoutesData
import com.route.routeme.R
import com.route.routeme.RouteArPath_v2
import com.route.routeme.databinding.FragmentFirstBinding
import com.route.viewmodel.RoutesDataViewModel
import io.github.g00fy2.quickie.QRResult
import io.github.g00fy2.quickie.ScanCustomCode
import io.github.g00fy2.quickie.ScanQRCode
import io.github.g00fy2.quickie.config.ScannerConfig
import io.reactivex.disposables.CompositeDisposable

/**
 * Created by Ashwani Kumar Singh on 09,February,2023.
 */
class FirstFragment:Fragment() {


    private val scanCustomCode = registerForActivityResult(ScanCustomCode(), ::listenBarCodeResult)
    private lateinit var binding: FragmentFirstBinding
    private lateinit var model: RoutesDataViewModel

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentFirstBinding.inflate(inflater, container, false)
        return binding.root
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        model = ViewModelProvider(requireActivity())[RoutesDataViewModel::class.java]
        model.clearDisposable()


//        val navController = NavHostFragment.findNavController(this@FirstFragment)
//        navController.navigate(R.id.action_FirstFragment_to_DocumentFragment, bundleOf("Url" to "https://rtme.pl/TgtHills02") )
//        binding.title.text = "List of Routes"


        scanCustomCode.launch(
            ScannerConfig.build {
            setOverlayStringRes(R.string.scan_barcode) // string resource used for the scanner overlay
            setHapticSuccessFeedback(false) // enable (default) or disable haptic feedback when a barcode was detected
            setShowTorchToggle(false) // show or hide (default) torch/flashlight toggle button
            setShowCloseButton(true) // show or hide (default) close button
            setHorizontalFrameRatio(1.0f) // set the horizontal overlay ratio (default is 1 / square frame)
            setUseFrontCamera(false) // use the front camera
        }
        )

        binding.barcodeBtn.setOnClickListener{
            scanCustomCode.launch(
                ScannerConfig.build {
                    setOverlayStringRes(R.string.scan_barcode) // string resource used for the scanner overlay
                    setHapticSuccessFeedback(false) // enable (default) or disable haptic feedback when a barcode was detected
                    setShowTorchToggle(false) // show or hide (default) torch/flashlight toggle button
                    setShowCloseButton(true) // show or hide (default) close button
                    setHorizontalFrameRatio(1.0f) // set the horizontal overlay ratio (default is 1 / square frame)
                    setUseFrontCamera(false) // use the front camera
                }
            )
        }

        binding.databaseBtn.setOnClickListener {
            model.loadDatabaseList()
        }

        binding.collectionBtn.setOnClickListener { model.loadRoutesMeDataCollection() }

        binding.documentBtn.setOnClickListener {
            //model.loadRoutesDocument();
            model.loadRoutesDocumentJson()
        }

        model.routesDocument.observe(requireActivity()) { routesData: RoutesData ->
            // update UI
            val routes = routesData.documents
            binding.resText.text = routes[0].id
        }

        model.routesDocumentJson.observe(
            requireActivity()
        ) { routesJsonData: JsonElement ->
            // update UI
            if (binding != null && binding.resText != null) {
                binding.resText.text = routesJsonData.toString()
            }
        }

        model.collection.observe(
            requireActivity()
        ) { collectionJson: JsonElement ->
            // update UI
            binding.resText.text = collectionJson.toString()
        }

        model.databaseList.observe(
            requireActivity()
        ) { databaseJson: JsonElement ->
            // update UI
            binding.resText.text = databaseJson.toString()
        }

        model.routesError.observe(
            requireActivity()
        ) { error: String? ->
            binding.resText.text = error
        }

    }

    override fun onResume() {
        super.onResume()
        binding.title.text = "Scan The Bar Code"

    }

    private fun listenBarCodeResult(result: QRResult) {
        when (result) {
            is QRResult.QRSuccess -> {
                var requiredValue = result.content.rawValue
                val navController = NavHostFragment.findNavController(this@FirstFragment)
                navController.navigate(R.id.action_FirstFragment_to_DocumentFragment, bundleOf("Url" to requiredValue) )
                binding.title.text = "List of Routes"

            }
            QRResult.QRUserCanceled -> {
                requireActivity().finish()
                showSnackBar("User canceled")
            }
            QRResult.QRMissingPermission -> {
                showSnackBar("Missing permission")
            }

            is QRResult.QRError -> {
                val msg = "${result.exception.javaClass.simpleName}: ${result.exception.localizedMessage}"
                showSnackBar(msg)
            }
        }

    }

    private fun showSnackBar(msg: String) {
        Snackbar.make(binding.root, msg, Snackbar.LENGTH_LONG).apply {
            view.findViewById<TextView>(com.google.android.material.R.id.snackbar_text)?.run {
                maxLines = 5
                setTextIsSelectable(true)
            }
        }.show()
    }

}